bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java

```diff
@@ -2542,9 +2543,16 @@ public State getState() {
             if (baseItem instanceof NumberItem numberItem) {
                 Unit<?> unit = numberItem.getUnit();
                 if (unit != null) {
-                    DecimalType decimalState = state.as(DecimalType.class);
-                    if (decimalState != null) {
-                        state = new QuantityType<>(decimalState, unit);
+                    if (state instanceof QuantityType<?> qtState) {
+                        qtState = qtState.toUnit(unit);
+                        if (qtState != null) {
+                            state = qtState;
+                        }
+                    } else {
+                        DecimalType decimalState = state.as(DecimalType.class);
+                        if (decimalState != null) {
+                            state = new QuantityType<>(decimalState, unit);
+                        }
                     }
```

Comment:
I'm really not sure about this one. The persistence service should return the correct type, i.e. a `QuantityType` if the state was originally a `QuantityType`. Which persistence service is not behaving like that?

Reply:
As far as I could tell from the code: JPA and MongoDB (for GroupItems).

Reply:
IMO this is a bug and should be fixed in the add-ons.

Reply:
Even when the service returns QuantityType, in theory, the unit could be different for different datapoints. If one wants to calculate with it, it needs to be converted to a common unit. It probably does not happen at the moment (most don’t even store the unit and just get it from the item). But it could happen with e.g. inmemory persistence.